### PR TITLE
19 data persistence

### PR DESCRIPTION
**Pull request by scissorkhizer**
This is the current implementation of data persistence within our program, there are some concerns with respect to clean and code smells but the program is functional! Please take a careful and thorough look on what else needs to be done before final submission.

### CODE REVIEW

**Review by missionStrawberry — DISMISSED** [review]
Looks good for now! testing should probably be split up into multiple tests if possible in a future issue.

**Review by eatingcorn — APPROVED** [review]
Looks good for now

File: src/main/java/screens/WeekViewScreen.java [hunk 1]
```diff
@@ -269,6 +271,11 @@ public void windowClosing(WindowEvent e) {
 
         if (confirmed == JOptionPane.YES_OPTION) {
             this.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
+            try {
+                WeekDataAccess.writeObject(WeekDataAccess.getDays());
+            } catch (IOException ex) {
+                throw new RuntimeException(ex);
```
Comment (Fard-Faru) [inline src/main/java/screens/WeekViewScreen.java:277]: please dont throw errors

**Review by Fard-Faru — CHANGES_REQUESTED** [review]

File: src/test/java/data_access/DataPersistenceTest.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+//
+//        assertFalse(WeekDataAccess.days.get(1).getEvents().containsKey("MAT237"));
+//        assertFalse(WeekDataAccess.days.get(2).getEvents().containsKey("Mat237"));
+//        assertFalse(WeekDataAccess.days.get(3).getEvents().containsKey("Csc236"));
+//        assertFalse(WeekDataAccess.days.get(4).getEvents().containsKey("Mat237"));
+//        assertFalse(WeekDataAccess.days.get(5).getTasks().containsKey("Feed dog"));
+//        assertFalse(WeekDataAccess.days.get(5).getEvents().containsKey("Csc236"));
+//        assertFalse(WeekDataAccess.days.get(1).getTasks().containsKey("Feed dog"));
+//        assertNotNull(WeekDataAccess.days);
+
```
Comment (Fard-Faru) [inline src/test/java/data_access/DataPersistenceTest.java:125]: are we still using these?

**Review by Fard-Faru — CHANGES_REQUESTED** [review]

**Review by Fard-Faru — CHANGES_REQUESTED** [review]

File: src/main/java/data_access/WeekDataAccess.java [hunk 3]
```diff
@@ -2,22 +2,23 @@
...
  */
 public class WeekDataAccess implements Serializable {
 
     /**
      * The ArrayList of days to be used for data persistence. The state of this data structure is saved
      * to a .ser file.
      */
-    protected static ArrayList<Day> days = new ArrayList<>();
+    protected static ArrayList<Day> days = new ArrayList<Day>();
+    private static final long serialVersionUID = 1L;
```
Comment (POPPININ) [inline src/main/java/data_access/WeekDataAccess.java]: Capitalize the variable name please.

File: src/main/java/data_access/WeekDataAccess.java [hunk 4]
```diff
@@ -41,14 +42,58 @@ public WeekDataAccess(ArrayList<Day> dataDays){
...
     }
 
     public static void setDays(ArrayList<Day> days) {
         WeekDataAccess.days = days;
     }
 
-    protected void save(){
+    /**
+     * Writes the contents of the days class variable onto a .ser file, persisting data
+     * @param days an ArrayList whose indices contain two hashmaps that contain events and tasks separately, representing the
```
Comment (POPPININ) [inline src/main/java/data_access/WeekDataAccess.java:46]: Technically the indices Day objects which contain hashmaps. Kindly correct this.

File: src/main/java/data_access/WeekDataAccess.java [hunk 5]
```diff
@@ -41,14 +42,58 @@ public WeekDataAccess(ArrayList<Day> dataDays){
...
+     *             week schedule
+     * @throws IOException
+     */
+    public static void writeObject(ArrayList<Day> days) throws IOException {
```
Comment (POPPININ) [inline src/main/java/data_access/WeekDataAccess.java]: Why does this method need to be public? It seems like it is only the save() method of this class which uses this method, so declaring it public doesn't serve any purpose, if not directly harmful and a violation of SOLID principles.

File: src/main/java/data_access/WeekDataAccess.java [hunk 6]
```diff
@@ -41,14 +42,58 @@ public WeekDataAccess(ArrayList<Day> dataDays){
...
+        objectOutputStream.flush();
+        objectOutputStream.close();
+    }
+
+    /**
+     * Reads the object from the .ser file and reinitialized day
+     * @throws ClassNotFoundException
+     * @throws IOException
+     */
+    public static void readObject() throws ClassNotFoundException, IOException {
```
Comment (POPPININ) [inline src/main/java/data_access/WeekDataAccess.java:63]: Again, why is this method public?

File: src/main/java/screens/WeekViewScreen.java [hunk 7]
```diff
@@ -269,6 +271,11 @@ public void windowClosing(WindowEvent e) {
 
         if (confirmed == JOptionPane.YES_OPTION) {
             this.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
+            try {
```
Comment (POPPININ) [inline src/main/java/screens/WeekViewScreen.java]: Please remove this. Only the data access classes of our use cases should invoke the save() method of WeekDataAccess.

**Review by Fard-Faru — APPROVED** [review]
Good Job Khizer

File: src/main/java/data_access/WeekDataAccess.java [hunk 8]
```diff
@@ -41,14 +42,58 @@ public WeekDataAccess(ArrayList<Day> dataDays){
...
+        File file = new File("CleanCalendarStorage.txt");
+        if (file.length() != 0) {//program has been initialized, assuming some content has been added
+            FileInputStream fileInputStream = new FileInputStream("CleanCalendarStorage.txt");
+            ObjectInputStream objectInputStream = new ObjectInputStream(fileInputStream);
+            WeekDataAccess.days = (ArrayList<Day>) objectInputStream.readObject();
+            objectInputStream.close();
+        }
+        else { //initializing program for the very first time, so file days with empty days
```
Comment (POPPININ) [inline src/main/java/data_access/WeekDataAccess.java:71]: The in-line comment looks odd.

**Review by eatingcorn — APPROVED** [review]
Looks good.

**Review by POPPININ — CHANGES_REQUESTED** [review]
Hi Khizer. Please address the comments that I have left.

File: src/main/java/screens/WeekViewScreen.java (hunk 7 above)
Comment (scissorkhizer) [reply src/main/java/screens/WeekViewScreen.java]: Reviewed an outdated commit.

File: src/main/java/data_access/WeekDataAccess.java (hunk 6 above)
Comment (scissorkhizer) [reply src/main/java/data_access/WeekDataAccess.java:63]: This is called in the beginning of main, which is vital to load in days from previous uses of the program.

File: src/main/java/data_access/WeekDataAccess.java (hunk 5 above)
Comment (scissorkhizer) [reply src/main/java/data_access/WeekDataAccess.java]: Review on outdated commit.

File: src/main/java/data_access/WeekDataAccess.java (hunk 3 above)
Comment (scissorkhizer) [reply src/main/java/data_access/WeekDataAccess.java]: review on outdated commit.

File: src/main/java/data_access/WeekDataAccess.java (hunk 8 above)
Comment (scissorkhizer) [reply src/main/java/data_access/WeekDataAccess.java:71]: Removed.

**Review by Fard-Faru — APPROVED** [review]

**Review by POPPININ — APPROVED** [review]
Looks good!

**Review by eatingcorn — APPROVED** [review]

**Pull request merged**